code/trans.c

```diff
@@ -0,0 +1,377 @@
+/* trans.c: TRANSFORMS IMPLEMENTATION
+ *
+ * $Id$
+ * Copyright 2011-2022 Ravenbrook Limited.  See end of file for license.
+ *
+ * A transform is a special kind of garbage collection that replaces references
+ * to a set of objects.  The transform is piggybacked onto a garbage
+ * collection by overriding the fix method for a trace.  The mapping used to
+ * replace the references is built up in a hash table by the client.
+ *
+ *
+ * Rationale
+ *
+ * This design was arrived at after some pain.  The MPS isn't really designed
+ * for this kind of thing, and the pools generally assume that they're doing
+ * a garbage collection when they're asked to condemn, scan, fix, and reclaim
+ * stuff.  This makes it very hard to apply the transform without also doing
+ * a garbage collection.  Changing this would require a significant reworking
+ * of the MPS to generalise its ideas, and would bloat the pool classes.
+ *
+ *
+ * Assumptions:
+ *
+ *   - Single-threaded mutator.  In fact this code might work if other mutator
+ *     threads are running, since the shield ought to operate correctly.
+ *     However, in this case there can only be one trace running so that the
+ *     correct fix method is chosen by ScanStateInit.
```

Comment:
m. This should be moved to design, investigated for accuracy, the relevant code connected to the design. rule.generic.self, rule.code.design, rule.code.justified

Reply:
nM. If this is a genuine restriction, it must be documented in the manual. rule.generic.achieve

Reply:
> m. This should be moved to design, investigated for accuracy, the relevant code connected to the design. rule.generic.self, rule.code.design, rule.code.justified

Fixed: Moved and cross-referenced in ebe0af12c.